File: src/main/java/random/telegramhomebot/services/HostService.java
package random.telegramhomebot.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.collections4.CollectionUtils;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import random.telegramhomebot.model.Host;
import random.telegramhomebot.model.HostState;
import random.telegramhomebot.model.HostTimeLog;
import random.telegramhomebot.repository.HostRepository;
import random.telegramhomebot.repository.HostTimeLogRepository;

import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static random.telegramhomebot.AppConstants.DATE_TIME_FORMATTER;
import static random.telegramhomebot.utils.NetUtils.comparingByIp;

@Slf4j
@RequiredArgsConstructor
@Service
public class HostService {

    public static final SimpleDateFormat TIME_DATE_FORMAT = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss");
    private final HostRepository hostRepository;
    private final HostTimeLogRepository hostTimeLogRepository;
    private final CommandRunnerService commandRunnerService;

    public List<Host> getAllHosts() {
        return hostRepository.findAll();
    }

    public Page<Host> getAllHosts(PageRequest pageable) {
        return hostRepository.findAll(pageable);
    }

    public void saveAllHosts(List<Host> hosts) {
        hostRepository.saveAll(hosts);
//		hostRepository.flush();
    }

    public Optional<Host> getHostByMac(String mac) {
        return hostRepository.findHostByMac(mac);
    }

    public Optional<Host> getHostById(UUID id) {
        return hostRepository.findById(id);
    }

    public void deleteHostById(UUID id) {
        hostRepository.deleteById(id);
    }

    public void saveHost(Host host) {
        hostRepository.save(host);
    }

    public List<Host> getReachableHosts() {
        return hostRepository.findReachableHosts();
    }

    public List<Host> getNewHosts(List<Host> storedHosts, List<Host> currentHosts) {
        String newHostNameStub = String.format("[NEW DEVICE] %s", LocalDateTime.now().format(DATE_TIME_FORMATTER));
        return currentHosts.stream()
                .filter(currentHost -> storedHosts.stream().noneMatch(currentHost::equals))
                .peek(host -> host.setDeviceName(newHostNameStub))
                .sorted(comparingByIp())
                .collect(Collectors.toList());
    }

    public List<Host> getHostsThatBecameReachable(List<Host> storedHosts, List<Host> currentHosts) {
        return currentHosts.stream()
                .filter(currentHost -> isReachable(storedHosts, currentHost))
                .distinct()
                .sorted(comparingByIp())
                .collect(Collectors.toList());
    }

    public List<Host> getHostsThatBecameNotReachable(List<Host> storedHosts, List<Host> currentHosts) {
        Stream<Host> hostsFailedStream1 = storedHosts.stream()
                .filter(storedHost -> reachableBecameNotFound(currentHosts, storedHost))
                .peek(host -> host.setState(HostState.FAILED));
        Stream<Host> hostsFailedStream2 = currentHosts.stream()
                .filter(currentHost -> reachableBecameNotReachable(storedHosts, currentHost));
        return Stream.concat(hostsFailedStream1, hostsFailedStream2)
                .sorted(comparingByIp())
                .collect(Collectors.toList());
    }

    public void saveTimeLogForHosts(List<Host> hosts) {
        if (CollectionUtils.isEmpty(hosts)) {
            return;
        }
        hostTimeLogRepository.saveAll(hosts.stream()
                .map(h -> new HostTimeLog(h, h.getState()))
                .collect(Collectors.toList()));
    }

    public void pingStoredHosts() {
        log.debug("Ping stored hosts...");
        List<Host> storedHosts = getAllHosts();
        log.debug("Stored hosts: \n{}", storedHosts);
        commandRunnerService.pingHosts(storedHosts);
    }

    public List<HostTimeLog> getLastHostTimeLogs(int logCount) {
        Pageable page = PageRequest.of(0, logCount, Sort.Direction.DESC, "createdDate");
        return hostTimeLogRepository.findAll(page).stream()
                .sorted(Comparator.comparing(HostTimeLog::getCreatedDate))
                .collect(Collectors.toList());
    }

    public List<HostTimeLog> getLastHostTimeLogsForHost(Host host, int logCount) {
        Pageable page = PageRequest.of(0, logCount, Sort.Direction.DESC, "createdDate");
        return hostTimeLogRepository.findHostTimeLogByHost(page, host).stream()
                .sorted(Comparator.comparing(HostTimeLog::getCreatedDate))
                .collect(Collectors.toList());
    }

    public String getLastHostTimeLogsAsString(int logCount) {
        return getLastHostTimeLogs(logCount).stream()
                .map(this::convertTimeLog)
                .collect(Collectors.joining("\n"));
    }

    private String convertTimeLog(HostTimeLog log) {
        return TIME_DATE_FORMAT.format(log.getCreatedDate())
               + "\t" + log.getState() + "\t\t" + log.getHost().getDeviceName();
    }

    private boolean isReachable(List<Host> storedHosts, Host currentHost) {
        return !HostState.FAILED.equals(currentHost.getState())
               && storedHosts.stream()
                       .anyMatch(storedHost -> storedHost.equals(currentHost)
                                               && HostState.FAILED.equals(storedHost.getState()));
    }

    private boolean reachableBecameNotFound(List<Host> currentHosts, Host storedHost) {
        return !HostState.FAILED.equals(storedHost.getState()) && currentHosts.stream().noneMatch(storedHost::equals);
    }

    private boolean reachableBecameNotReachable(List<Host> storedHosts, Host currentHost) {
        return HostState.FAILED.equals(currentHost.getState())
               && storedHosts.stream()
                       .anyMatch(storedHost -> storedHost.equals(currentHost)
                                               && !HostState.FAILED.equals(storedHost.getState()));
    }
}

File: src/main/java/random/telegramhomebot/services/HostService.kt
package random.telegramhomebot.services

import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import random.telegramhomebot.AppConstants
import random.telegramhomebot.model.Host
import random.telegramhomebot.model.HostState
import random.telegramhomebot.model.HostTimeLog
import random.telegramhomebot.repository.HostRepository
import random.telegramhomebot.repository.HostTimeLogRepository
import random.telegramhomebot.utils.NetUtils
import random.telegramhomebot.utils.logger
import java.text.SimpleDateFormat
import java.time.LocalDateTime
import java.util.*

@Service
class HostService(
    private val hostRepository: HostRepository,
    private val hostTimeLogRepository: HostTimeLogRepository,
    private val commandRunnerService: CommandRunnerService
) {
    val log = logger()

    fun getAllHosts(): List<Host> = hostRepository.findAll()
    fun getAllHosts(pageable: PageRequest): Page<Host?> = hostRepository.findAll(pageable)

    fun saveAllHosts(hosts: List<Host>) {
        hostRepository.saveAll(hosts)
        //		hostRepository.flush();
    }

    fun getHostByMac(mac: String?): Host? = hostRepository.findHostByMac(mac)
    fun getHostById(id: UUID): Optional<Host?> = hostRepository.findById(id)
    fun deleteHostById(id: UUID) = hostRepository.deleteById(id)
    fun saveHost(host: Host): Host = hostRepository.save(host)
    fun getReachableHosts(): List<Host> = hostRepository.findReachableHosts()
    fun getWakeOnLanEnableHosts(): List<Host> = hostRepository.findWakeOnLanEnabledHosts()
    fun count() = hostRepository.count()

    fun getNewHosts(storedHosts: List<Host?>, currentHosts: List<Host>): List<Host> {
        val newHostNameStub =
            String.format("[NEW DEVICE] %s", LocalDateTime.now().format(AppConstants.DATE_TIME_FORMATTER))
        return currentHosts
            .filter { currentHost -> storedHosts.none { other -> currentHost == other } }
            .onEach { it.deviceName = newHostNameStub }
            .sortedWith(NetUtils.comparingByIp())
    }

    fun getHostsThatBecameReachable(storedHosts: List<Host>, currentHosts: List<Host>): List<Host> {
        return currentHosts
            .filter { isReachable(storedHosts, it) }
            .distinct()
            .sortedWith(NetUtils.comparingByIp())
    }

    fun getHostsThatBecameNotReachable(storedHosts: List<Host>, currentHosts: List<Host>): List<Host> {
        val hostsFailedStream1 = storedHosts
            .filter { reachableBecameNotFound(currentHosts, it) }
            .onEach { it.state = HostState.FAILED }
        val hostsFailedStream2 = currentHosts.filter { reachableBecameNotReachable(storedHosts, it) }
        return hostsFailedStream1 + hostsFailedStream2
    }

    fun saveTimeLogForHosts(hosts: List<Host>) {
        if (hosts.isEmpty()) {
            return
        }
        val timeLogs = hosts.map { HostTimeLog(it, it.state!!) }
        hostTimeLogRepository.saveAll(timeLogs)
    }

    fun pingStoredHosts() {
        log.debug("Ping stored hosts...")
        val storedHosts = getAllHosts()
        log.debug("Stored hosts: \n{}", storedHosts)
        commandRunnerService.pingHosts(storedHosts)
    }

    private fun getLastHostTimeLogs(logCount: Int): List<HostTimeLog> {
        val page: Pageable = PageRequest.of(0, logCount, Sort.Direction.DESC, "createdDate")
        return hostTimeLogRepository.findAll(page).sortedWith(Comparator.comparing(HostTimeLog::createdDate))
    }

    fun getLastHostTimeLogsForHost(host: Host?, logCount: Int): List<HostTimeLog> {
        val page: Pageable = PageRequest.of(0, logCount, Sort.Direction.DESC, "createdDate")
        return hostTimeLogRepository.findHostTimeLogByHost(page, host)
            .sortedWith(Comparator.comparing(HostTimeLog::createdDate))
    }

    fun getLastHostTimeLogsAsString(logCount: Int): String =
        getLastHostTimeLogs(logCount).joinToString(separator = "\n") { convertTimeLog(it) }

    private fun convertTimeLog(log: HostTimeLog) =
        "${TIME_DATE_FORMAT.format(log.createdDate)} \t ${log.state} \t\t ${log.host.deviceName}"

    private fun isReachable(storedHosts: List<Host>, currentHost: Host) =
        HostState.FAILED != currentHost.state
                && storedHosts.any { storedHost -> storedHost == currentHost && HostState.FAILED == storedHost.state }

    private fun reachableBecameNotFound(currentHosts: List<Host>, storedHost: Host) =
        HostState.FAILED != storedHost.state && currentHosts.none { storedHost == it }

    private fun reachableBecameNotReachable(storedHosts: List<Host>, currentHost: Host) =
        HostState.FAILED == currentHost.state && storedHosts.any { it == currentHost && HostState.FAILED != it.state }

    companion object {
        private val TIME_DATE_FORMAT = SimpleDateFormat("dd/MM/yyyy HH:mm:ss")
    }
}